### ISSUE TITLE
Dataset ctx_left/ctx_right extension: ctx_clip_to_valid option

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn#1564]
When you have some feature extraction or downsampling module inside your network with padding=valid (standard stft is also padding=valid), it means that for a kernel size K, K-1 frames left/right will be gone. Ignoring the downsampling aspect, you have:

$T_{\textrm{in}} - (K - 1) = T_{\textrm{out}}$

Now, with chunking, you would get:

$T_{\textrm{in}} / N - (K - 1) \ne T_{\textrm{out}} / N$

Using ctx_left/ctx_right cannot fix this. Assuming $K$ uneven, if you set ctx_left=ctx_right=(K-1)/2, then you get:

$T_{\textrm{in}} / N + (K - 1) - (K - 1) = T_{\textrm{in}} / N \ne T_{\textrm{out}} / N$

Additionally to ctx_left/ctx_right, you want to remove some context on the sequence level, not on chunk level, such that you have:

$(T_{\textrm{in}} - (K - 1)) / N = T_{\textrm{out}} / N$

We (@robin-p-schmitt) thought about adding some `global_ctx_left`/`global_ctx_right`, and you would set global_ctx_left=global_ctx_right=-(K-1)/2 for this example.

But then we realized, this effectively means that at the left boundary on sequence level and on the right boundary on sequence level, the local (chunk level) ctx and global ctx would cancel out. Currently we have this code:
```python
for seq_idx, t_start, t_end in self.iterate_seqs(recurrent_net=recurrent_net, used_data_keys=used_data_keys):
    ...
    t_start -= self.ctx_left
    t_end += self.ctx_right
```
For this example, we could also introduce a `ctx_clip_to_valid` option which would clip `t_start`/`t_end` to valid frames, like so:
```python
if ctx_clip_to_valid:
    t_start = max(t_start, 0)
    t_end = min(t_end, T)  # t_end is exclusive
```
`ctx_clip_to_valid` is conceptually a bit simpler than global_ctx_left/global_ctx_right, and solves at least our problem, so we suggest to go with this simpler solution.

Note, the code I posted above is currently only used for TF. For PyTorch, we have the independent `ChunkingIterDataPipe` implementation, which even does not currently support ctx_left/ctx_right. So I would suggest to add `ctx_left`/`ctx_right` and `ctx_clip_to_valid` there.

cc @michelwi, @JackTemaki, @mmz33, @Icemole, @kuacakuaca

**Comment by michelwi** [comment]
I "solved" this by making the chunk size in one dimension a bit larger than the chunk size of the other. e.g.
```
chunking = ({"classes": 500, "data": 80240}, {"classes": 250, "data": 40000})
min_chunk_size = {"classes": 1, "data": 241}
```
so that after feature extraction, which has a window_size of 400 (25ms) and a window_shift of 160 (10ms) I always get the exact same number of features as there are classes.

**Comment by albertz** [comment]
Your feature extraction is inside the network/model? Does it use padding=valid?

So with padding=valid, what you get is (80240-400+1)/160=499.00625, i.e. 500 frames (ceildiv), if the chunk is complete. But it might also be less.

In our suggestion, you would add ctx_left=199, ctx_right=200. So it gets up to 80399 frames per chunk on data. But with ctx_clip_to_valid, for the first/last chunk, it would be effectively less.

I'm not sure the downsampling plays a role. So without it, let's assume a kernel size of 3 for your example. So you set:
```
chunking = ({"classes": 500, "data": 502}, {"classes": 250, "data": 250})
min_chunk_size = {"classes": 1, "data": 3}
```
And for the total seq len, you always would have $T_{\textrm{in}} - 2 = T_{\textrm{out}}$. But you have the same now for every chunk. So it should be fine.

Actually, with what we proposed here, the ctx_clip_to_valid=True and ctx_left=ctx_right=1, we would not have the same for every chunk. The first chunk would effectively have less frames. E.g. for:
```
chunking = ({"classes": 500, "data": 500}, {"classes": 250, "data": 250})
ctx_left = ctx_right = 1
ctx_clip_to_valid = True
```
you could get a chunk of size data:501,classes:500. So what we proposed here does not work.

I guess I can close this then.

**Comment by michelwi** [comment]
> Your feature extraction is inside the network/model? Does it use padding=valid?

I think it is more like same padding, but the padding is custom as implemented in [i6_models.primitives.feature_extraction](https://github.com/rwth-i6/i6_models/blob/ef22941079557341c26883766da06a9eae55a96d/i6_models/primitives/feature_extraction.py#L190) or [tf.signal.stft](https://www.tensorflow.org/api_docs/python/tf/signal/stft) with `pad_end=True`

**Comment by albertz** [comment]
> > Your feature extraction is inside the network/model? Does it use padding=valid?
> 
> I think it is more like same padding, but the padding is custom as implemented in [i6_models.primitives.feature_extraction](https://github.com/rwth-i6/i6_models/blob/ef22941079557341c26883766da06a9eae55a96d/i6_models/primitives/feature_extraction.py#L190) 

I wonder a bit whether this formula is really correct? This looks really custom. With standard padding=valid, you would have:

ceildiv(T - win_length + 1, hop_length)

With standard padding=same, you would have:

ceildiv(T, hop_length)

However, in this code, (ignoring the max) we have:

(T - win_length + hop_length - 1) // hop_length + 1 = ceildiv(T - win_length, hop_length) + 1 = ceildiv(T - win_length + hop_length, hop_length)

So, e.g. for hop_length=10, win_length=20. For T=20, we get 1, for T=21,...,30, we get 2, for T=31, we get 3.
With padding=valid instead: for T=20,...,29, we get 1, for T=30,...,39, we get 2.
Or with padding=same instead: for T=1,...,10, we get 1, for T=11,...,20, we get 2, for T=21,...,30, we get 3.

All three are different. To get the same as this code, you need to add hop_length-1 padding frames. (But e.g. for Torch `Conv1d` or so, it would actually be `padding=(hop_length-1)//2` if hop_length is odd. Or you just do custom padding.)

Is this really what we do in RASR? So we pad exactly hop_length-1 frames? Why?

But in any case, I think the approach you described can also be done here, right?
* In case of padding=same (i.e. padding win_length-1 frames in total), no modification is needed for chunking (no extra frames to the chunk size).
* In case of padding=valid (i.e. padding 0 frames in total), you add win_length-1 frames in total to the chunk size.
* In general when the model pads model_pad_total frames in total, you add (win_length-1)-model_pad_total to the chunk size.
* So, here, with padding hop_length-1 frames in total, you would add (win_length-1)-(hop_length-1) = win_length-hop_length frames to the chunk size.

> or [tf.signal.stft](https://www.tensorflow.org/api_docs/python/tf/signal/stft) with `pad_end=True`

In the RETURNN TF `StftLayer`, we use `pad_end=False`, i.e. like padding=valid. This is what I have seen for most of our feature extraction inside the network configs. In my RF-based setup (and also from @robin-p-schmitt), we also have the same (via `rf.stft`).

**Comment by michelwi** [comment]
> Is this really what we do in RASR? So we pad exactly hop_length-1 frames? Why?

I have no idea why it was done this way, but as I want to be exactly compatible with the number of frames that rasr produces.

> But in any case, I think the approach you described can also be done here, right?

yes exactly. My comment was just to answer your question and not to argue to reopen this issue or change the implementation.
